### ISSUE TITLE
[BUG] - Script validation passes with tx build-raw but not with tx build

### ISSUE THREAD

**Issue by tdiesler** [input-output-hk/cardano-node#3048]
**Internal/External**
*External*   otherwise.

**Area**
*Plutus*    Related to Plutus Scripts (Alonzo).

**Summary**

I have a script that does not validate when building the Tx with `cardano-cli transaction build` like this ...

```
TX_SCRIPT="dc17f12d92e6cd94279f84800933e6074d4c338373d30d2125d9f6c7fba06f10#1"

TX_IN1="3ada602358619cc701b7420d49cb1a7b7bd6e208a4ba7f1aff79c35261643cd4#1"

TX_COL="09a0e6c42682440dbd71cf1db0ddb5b35701d2dc8e8f2aee837db64e71b0ab85#0"

SEND_LVC=10000000

# Build, sign and submit the transaction
cardano-cli transaction build \
  --alonzo-era \
  --testnet-magic 8 \
  --tx-in $TX_IN1 \
  --tx-in $TX_SCRIPT \
  --tx-in-script-file /var/cardano/local/scripts/helloworld5a.plutus \
  --tx-in-datum-value 0 \
  --tx-in-redeemer-file /var/cardano/local/scratch/redeemer-secret2.json \
  --tx-in-collateral $TX_COL \
  --tx-out $TO_ADDR+$SEND_LVC \
  --change-address $PAYMENT_ADDR \
  --protocol-params-file /var/cardano/local/scratch/protocol.json \
  --out-file /var/cardano/local/scratch/tx.raw \
&& cardano-cli transaction sign \
  --tx-body-file /var/cardano/local/scratch/tx.raw \
  --signing-key-file /var/cardano/local/keys/alonzo/wallets/wallet1.skey \
  --out-file /var/cardano/local/scratch/tx.signed \
&& cardano-cli transaction submit \
  --tx-file /var/cardano/local/scratch/tx.signed \
  --testnet-magic 8
```

but it does validate using the same set of UTxO when building the Tx with `cardano-cli transaction build-raw` like this 

```
TX_SCRIPT="dc17f12d92e6cd94279f84800933e6074d4c338373d30d2125d9f6c7fba06f10#1"
TX_SCRIPT_LVC="10000000"

TX_IN1="3ada602358619cc701b7420d49cb1a7b7bd6e208a4ba7f1aff79c35261643cd4#1"
TX_IN1_LVC="1117521600"

TX_COL="09a0e6c42682440dbd71cf1db0ddb5b35701d2dc8e8f2aee837db64e71b0ab85#0"
TX_COL_LVC="2000000000"

# Calculate the change to send back to PAYMENT_ADDR
ExCPU=29392000
ExMem=7280
ExFct=30
UNITS="($(($ExFct*$ExCPU)),$(($ExFct*$ExMem)))"
FEES_LVC=$(($ExFct * ($ExCPU + $ExMem) + 500000))
REFUND_LVC=$(($TX_IN1_LVC - $FEES_LVC))
SEND_LVC=$TX_SCRIPT_LVC
echo "Send=$SEND_LVC, Refund=$REFUND_LVC, Fees=$FEES_LVC, Units=$UNITS"

# Build, sign and submit the transaction
cardano-cli transaction build-raw \
  --alonzo-era \
  --fee $FEES_LVC \
  --tx-in $TX_IN1 \
  --tx-in $TX_SCRIPT \
  --tx-in-script-file /var/cardano/local/scripts/helloworld5a.plutus \
  --tx-in-datum-value 0 \
  --tx-in-redeemer-file /var/cardano/local/scratch/redeemer-secret2.json \
  --tx-in-execution-units "$UNITS" \
  --tx-in-collateral $TX_COL \
  --tx-out $TO_ADDR+$SEND_LVC \
  --tx-out $PAYMENT_ADDR+$REFUND_LVC \
  --protocol-params-file /var/cardano/local/scratch/protocol.json \
  --out-file /var/cardano/local/scratch/tx.raw \
&& cardano-cli transaction sign \
  --tx-body-file /var/cardano/local/scratch/tx.raw \
  --signing-key-file /var/cardano/local/keys/alonzo/wallets/wallet1.skey \
  --out-file /var/cardano/local/scratch/tx.signed \
&& cardano-cli transaction submit \
  --tx-file /var/cardano/local/scratch/tx.signed \
  --testnet-magic 8
```

https://github.com/tdiesler/nessus-cardano/blob/alonzo/plutus/alonzo/4_Alonzo-purple-exercise-4.md#5a2-unlock-the-funds-in-the-script

**Comment by james-iohk** [comment]
So it passed validation of `transaction build` and a txbody is produced, what's the error `submit`?

**Comment by tdiesler** [comment]
It failed at build, with didn't validate. Let me try again ...

**Comment by james-iohk** [comment]
Just to clarify, `build-raw` doesn't run validation. Does that signed tx successfully `submit`?

**Comment by tdiesler** [comment]
Yes, it submits the Tx, which then unlock the the funds from the script successfully.
It seems, that the error `tx build` runs into does not happen with the `tx build-raw` case.

**Comment by james-iohk** [comment]
Ok thanks. Would you mind providing source for `helloworld5a.plutus`, or is it the same as [HelloWorld.hs](https://github.com/input-output-hk/Alonzo-testnet/blob/main/resources/plutus-sources/plutus-helloworld/src/Cardano/PlutusExample/HelloWorld.hs)? (I suspect it is different because your datum is `0`)

**Comment by tdiesler** [comment]
It's all in the repo. Here is the [entrypoint](https://github.com/tdiesler/nessus-cardano/blob/alonzo/plutus/alonzo/4_Alonzo-purple-exercise-4.md#5a1-lock-some-funds-in-the-script) for this 

```
cd ~/git/nessus-cardano/plutus/alonzo/plutus-sources/plutus-helloworld \
  && cabal run plutus-helloworld5a 1 helloworld5a.plutus \
  && cp helloworld5a.plutus ../../plutus-scripts \
  && mv helloworld5a.plutus ~/cardano/scripts \
  && cardano-cli address build \
    --payment-script-file /var/cardano/local/scripts/helloworld5a.plutus \
    --out-file /var/cardano/local/scratch/helloworld5a.addr \
    --testnet-magic 8 \
  && SCRIPT_ADDR=$(cat ~/cardano/scratch/helloworld5a.addr) \
  && echo "SCRIPT_ADDR=${SCRIPT_ADDR}"

ExBudget {_exBudgetCPU = ExCPU 29392000, _exBudgetMemory = ExMemory 7280}
SCRIPT_ADDR=addr_test1wp5aunhx9x7k0v86agmsfpcsm9w9vj6u9gwde4gxzk8a3gq95cqrw
```

The  validator script is [here](https://github.com/tdiesler/nessus-cardano/blob/alonzo/plutus/alonzo/plutus-sources/plutus-helloworld/src/Cardano/PlutusExample/HelloWorld5a.hs)

**Comment by james-iohk** [comment]
I see now. I wasn't aware that we could validate PubKeyHash from script context yet (mismatch with plutus and ledger). I'll look at this tomorrow. Meanwhile, what's the way you got your address pubkey, is it the base16 value?

**Comment by tdiesler** [comment]
That's documented here
https://github.com/tdiesler/nessus-cardano/blob/alonzo/plutus/alonzo/4_Alonzo-purple-exercise-4.md#5-set-up-new-payment-addresses-for-wallet1--wallet2

```
# Extract the PubKey Hashes
cardano-cli address key-hash \
  --payment-verification-key-file /var/cardano/local/keys/alonzo/wallets/wallet1.vkey \
  --out-file /var/cardano/local/scratch/wallet1-pubkey.hash \
&& cardano-cli address key-hash \
  --payment-verification-key-file /var/cardano/local/keys/alonzo/wallets/wallet2.vkey \
  --out-file /var/cardano/local/scratch/wallet2-pubkey.hash \
&& WALLET_HASH1="$(cat ~/cardano/scratch/wallet1-pubkey.hash)" \
&& WALLET_HASH2="$(cat ~/cardano/scratch/wallet2-pubkey.hash)" \
&& echo "WALLET_HASH1=$WALLET_HASH1" \
&& echo "WALLET_HASH2=$WALLET_HASH2"
```

**Comment by james-iohk** [comment]
To summarise what I did:
- Took a copy of [HelloWorld5a.hs](https://github.com/tdiesler/nessus-cardano/blob/c2924d82168069907d303230a16614223963e672/plutus/alonzo/plutus-sources/plutus-helloworld/src/Cardano/PlutusExample/HelloWorld5a.hs) and edited it to include my own pubkeyhash "5acb7060b154dcac9fe1367868288a2c9a4ffdb9e351f5151cd64f8a" for "secret1". (double checked that this is the only diff with original)
- Serialised the modified script, resulting address on alonzo-purple network is `addr_test1wzjx7ne5hy6gg8rc0sqm4plkmnxe9ylhqudpmnp59lt3m5sjf56ka`
- Sent funds to the script address with datum hash for value `0`
- Using `transaction build` with `redeemer-secret2.json` I proved the error occurs, as expected (because this is not my pubkeyhash)
- Using `transaction build` with `redeemer-secret1.json` I proved no eorrs occurs and I could then sign and submit that transaction to successfully unlock funds.
- I did not try anything with `transaction build-raw`

**Comment by tdiesler** [comment]
I reviewed this again and indeed there was a mixup with the secret value for the tx build case. 
I'm sorry for wasting your time.

**Comment by james-iohk** [comment]
That's ok Tom. We appreciate your coverage and feedback. I learnt the public key hash is working as expected, I'll let the moderators know!

**Comment by ashisherc** [comment]
@james-iohk @tdiesler Is it possible to validate only the script execution part without running a full node?

**Comment by tdiesler** [comment]
Yes, if `cardano-cli transaction build` succeeds locally, it will also succeed on-chain - at least that's the theory.

**Comment by ashisherc** [comment]
Yes, but it requires a socket connection to cardano-node. Whereas build-raw doesn't, but also does not do script verification.
